ktor-shared/ktor-openapi-schema/common/src/io/ktor/openapi/ReferenceOr.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2014-2025 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.openapi
+
+import kotlinx.serialization.*
+import kotlinx.serialization.builtins.serializer
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import kotlin.jvm.JvmInline
+
+private const val RefKey = $$"$ref"
+private const val RecursiveRefKey = $$"$recursiveRef"
+
+/**
+ * Defines Union [A] | [Reference]. A lot of types like Header, Schema, MediaType, etc. can be
+ * either a direct value or a reference to a definition.
+ */
+@Serializable(with = ReferenceOr.Companion.Serializer::class)
+public sealed interface ReferenceOr<out A> {
+    public data class Reference(
+        public val ref: String,
+        public val isRecursive: Boolean = false,
+    ) : ReferenceOr<Nothing>
+
+    @JvmInline public value class Value<A>(public val value: A) : ReferenceOr<A>
+
+    public fun valueOrNull(): A? =
+        when (this) {
+            is Reference -> null
+            is Value -> value
+        }
+
+    public companion object {
+        private const val schema: String = "#/components/schemas/"
+        private const val responses: String = "#/components/responses/"
+        private const val parameters: String = "#/components/parameters/"
+        private const val requestBodies: String = "#/components/requestBodies/"
+        private const val pathItems: String = "#/components/pathItems/"
+
+        public fun schema(name: String): Reference = Reference("$schema$name")
+
+        public fun <A> value(value: A): ReferenceOr<A> = Value(value)
+
+        internal class Serializer<T>(private val dataSerializer: KSerializer<T>) :
+            KSerializer<ReferenceOr<T>> {
+            @OptIn(InternalSerializationApi::class, ExperimentalSerializationApi::class)
+            override val descriptor: SerialDescriptor =
+                buildSerialDescriptor("io.ktor.openapi.ReferenceOr", SerialKind.CONTEXTUAL)
+            private val refDescriptor = buildClassSerialDescriptor("io.ktor.openapi.ReferenceOr.Reference") {
+                element<String>(RefKey)
+            }
+            private val recursiveRefDescriptor = buildClassSerialDescriptor("io.ktor.openapi.ReferenceOr.Reference") {
```

Comment:
Both `refDescriptor` and `recursiveRefDescriptor` use the same name `io.ktor.openapi.ReferenceOr.Reference`. Is it okay?